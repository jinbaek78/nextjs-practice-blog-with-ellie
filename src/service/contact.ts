import { EmailData } from './email';

export async function sendContactEmail(email: EmailData) {
  const response = await fetch('http://localhost:3000/api/contact', {
    method: 'POST',
    body: JSON.stringify(email),
    headers: {
      'Content-Type': 'application/json',
    },
  });

  const data = await response.json();
  if (!response.ok) {
    throw new Error(data.message || 'failed to request to server');
  }

  return data;
}
