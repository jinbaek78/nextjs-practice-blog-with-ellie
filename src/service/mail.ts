import { BannerData } from '@/components/Banner';
import { Form } from '@/components/ContactForm';
import { object, string, number, date, InferType } from 'yup';

export async function sendEmail(
  data: Form
  // callback: (banner: BannerData) => void
): Promise<BannerData> {
  let formSchema = object({
    subject: string().required(),
    from: string().email().required(),
    message: string().min(2).required(),
  });

  let dataValidationResult;
  try {
    dataValidationResult = await formSchema.validate(data);
  } catch (err) {
    console.log(err);
    return { message: 'please, write the right format', state: 'error' };
  }

  const formHeader = new Headers();
  formHeader.append('Content-Type', 'application/json');

  const requestOptions = {
    method: 'POST',
    headers: formHeader,
    body: JSON.stringify(data),
  };

  let res;
  let resultBanner: BannerData;
  try {
    res = await fetch('http://localhost:3000/api/contact', requestOptions);
  } catch (err) {
    console.log(err);
    return (resultBanner = { message: 'fail', state: 'error' });
  }

  return res.status === 200
    ? { message: 'success', state: 'success' }
    : { message: 'fail', state: 'error' };

  // return fetch('http://localhost:3000/api/contact', requestOptions)
  //   .then((response) => response.json())
  //   .then((result) => console.log(result))
  //   .catch((error) => console.log('error', error));
}
