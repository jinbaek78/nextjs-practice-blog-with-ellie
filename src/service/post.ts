import fs from 'fs/promises';
import path from 'path';
import { cwd } from 'process';

export type Post = {
  title: string;
  description: string;
  date: string;
  category: string;
  path: string;
  featured: boolean;
};

export async function getPostsData(): Promise<Post[]> {
  const postsData = await JSON.parse(
    await fs.readFile(`${path.join(cwd(), 'public', 'data', 'posts.json')}`, {
      encoding: 'utf-8',
    })
  );

  return postsData;
}
