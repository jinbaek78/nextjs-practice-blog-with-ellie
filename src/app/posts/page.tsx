import Posts from '@/components/Posts';
import { getAllPosts } from '@/service/posts';

export default async function PostsPage() {
  const posts = await getAllPosts();

  return <Posts posts={posts} />;
}
