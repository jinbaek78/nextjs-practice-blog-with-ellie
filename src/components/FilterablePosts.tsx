'use client';

import { Post } from '@/service/posts';
import { useState } from 'react';
import PostsGrid from './PostsGrid';
import Categories from './Categories';

const ALL_POSTS = 'All Posts';

type Props = {
  posts: Post[];
  categories: string[];
};
export default function FilterablePosts({ posts, categories }: Props) {
  console.log('FilterablePosts called');
  const [selected, setSelected] = useState(ALL_POSTS);
  const filtered =
    selected === ALL_POSTS
      ? posts
      : posts.filter((post) => post.category === selected);
  return (
    <section className="flex">
      <PostsGrid posts={filtered} />
      <Categories
        categories={[ALL_POSTS, ...categories]}
        selected={selected}
        onClick={setSelected}
      />
    </section>
  );
}
