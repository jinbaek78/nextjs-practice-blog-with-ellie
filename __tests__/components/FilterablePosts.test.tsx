import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import PostsGrid from '@/components/PostsGrid';
import Categories from '@/components/Categories';
import FilterablePosts from '@/components/FilterablePosts';
import { Post } from '@/service/posts';
import PostData from '../../public/data/posts.json';

jest.mock('@/components/PostsGrid');
// jest.mock('@/components/Categories');

describe('FilterablePosts', () => {
  const posts: Post[] = PostData;
  const categories = [...new Set(posts.map((post) => post.category))];

  afterEach(() => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockReset();
  });

  it('filters ALL Posts correctly', () => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    // (Categories as jest.MockedFunction<typeof Categories>).mockImplementation();
    render(<FilterablePosts posts={PostData} categories={categories} />);
    expect(
      (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[0][0]
    ).toEqual({ posts });
  });

  // should I test FilterablePosts and Categories to check the filter works welL?
  it('filters javascript Posts correctly', async () => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    render(<FilterablePosts posts={PostData} categories={categories} />);
    const filterOption = 'javascript';
    await fireEvent.click(screen.getByText(filterOption));
    const filtered = posts.filter((post) => post.category === filterOption);
    await waitFor(() => {
      expect(
        PostsGrid as jest.MockedFunction<typeof PostsGrid>
      ).toHaveBeenCalledTimes(2);
    });

    expect(
      (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[1][0]
        .posts
    ).toStrictEqual(filtered);

    // const { container } = render(
    //   <FilterablePosts posts={PostData} categories={categories} />
    // );
    // expect(container).toMatchSnapshot();

    // type Props = {
    //   onClick: (category: string) => void;
    // };
    // (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    // (Categories as jest.MockedFunction<typeof Categories>).mockImplementation(
    //   ({ onClick }: Props): JSX.Element => {
    //     setTimeout(() => {
    //       onClick('javascript');
    //     }, 100);
    //     return <h1>just JSX</h1>;
    //   }
    // );

    // render(<FilterablePosts posts={PostData} categories={categories} />);
    // const javascriptPosts = posts.filter(
    //   (post) => post.category === 'javascript'
    // );
    // console.log('javascriptPosts:', javascriptPosts);
    // await waitFor(() => {
    //   // console.log(
    //   //   (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[0][0]
    //   // );
    //   expect(
    //     (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[1][0]
    //   ).toEqual(1);
    // });
  });

  it('filters my story Posts correctly', async () => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    render(<FilterablePosts posts={PostData} categories={categories} />);
    const filterOption = 'my story';
    await fireEvent.click(screen.getByText(filterOption));
    const filtered = posts.filter((post) => post.category === filterOption);
    await waitFor(() => {
      expect(
        PostsGrid as jest.MockedFunction<typeof PostsGrid>
      ).toHaveBeenCalledTimes(2);
    });

    expect(
      (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[1][0]
        .posts
    ).toStrictEqual(filtered);
  });

  it('filters backend Posts correctly', async () => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    render(<FilterablePosts posts={PostData} categories={categories} />);
    const filterOption = 'backend';
    await fireEvent.click(screen.getByText(filterOption));
    const filtered = posts.filter((post) => post.category === filterOption);
    await waitFor(() => {
      expect(
        PostsGrid as jest.MockedFunction<typeof PostsGrid>
      ).toHaveBeenCalledTimes(2);
    });

    expect(
      (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[1][0]
        .posts
    ).toStrictEqual(filtered);
  });
  it('filters frontend Posts correctly', async () => {
    (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mockImplementation();
    render(<FilterablePosts posts={PostData} categories={categories} />);
    const filterOption = 'frontend';
    await fireEvent.click(screen.getByText(filterOption));
    const filtered = posts.filter((post) => post.category === filterOption);
    await waitFor(() => {
      expect(
        PostsGrid as jest.MockedFunction<typeof PostsGrid>
      ).toHaveBeenCalledTimes(2);
    });

    expect(
      (PostsGrid as jest.MockedFunction<typeof PostsGrid>).mock.calls[1][0]
        .posts
    ).toStrictEqual(filtered);
  });
});
