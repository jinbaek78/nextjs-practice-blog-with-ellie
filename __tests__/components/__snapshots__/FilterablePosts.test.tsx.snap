// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterablePosts filters other Posts correctly 1`] = `
<div>
  <section
    class="flex"
  >
    <section
      class="text-center p-4"
    >
      <h2
        class="text-lg font-bold border-b border-sky-500 mb-2"
      >
        Category
      </h2>
      <ul>
        <li
          class="cursor-pointer hover:text-sky-500 text-sky-600"
        >
          All Posts
        </li>
        <li
          class="cursor-pointer hover:text-sky-500 false"
        >
          my story
        </li>
        <li
          class="cursor-pointer hover:text-sky-500 false"
        >
          javascript
        </li>
        <li
          class="cursor-pointer hover:text-sky-500 false"
        >
          backend
        </li>
        <li
          class="cursor-pointer hover:text-sky-500 false"
        >
          frontend
        </li>
      </ul>
    </section>
  </section>
</div>
`;
